### PR TITLE
Persist and restore both contracts' `origin` and `head`

### PR DESCRIPTION
**Pull request by ZER0**
- Update rust toolchain
- Add `NetworkStateId` structure to handle both origin/head `PersistedId`
- Add utility methods to `NetworkStateId` to save / load from file
- Change `NetworkState::persist` to store both origin/head
- Add `persist` module under `state`
- Change `NetworkState::persist()` to accept the same type of
  `Persistence::persist()`

  Having as input type `&BackendCtor<B>` instead of the function pointer,
  semplify greatly the logic and the tests, since the temp_folder won't
  be dropped immediately.
- Change `NetworkState::restore` to restore both origin/head

- Update integration tests

   Having now the `&BackendCtor<B>` for `Network::persist` simplify the
   tests, removing the unnecessary manual "teardown" to clean up

- Bump to `0.8.0-rc.0`

Resolves: #289

### CODE REVIEW

File: src/state/persist.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        let mut source = Source::new(&buf[..]);
+        let id = NetworkStateId::decode(&mut source)
+            .map_err(VMError::from_store_error)?;
+
+        Ok(id)
+    }
+
+    /// Write to the given path a [`NetworkStateId`]
+    pub fn write<P: AsRef<Path>>(&self, path: P) -> Result<(), VMError> {
+        let mut buf = [0u8; 36 * 2];
```
Comment (krl) [inline src/state/persist.rs:39]: using non-obvious numbers should be explained

**Review by krl — APPROVED** [review]
With the commenting nit, this look reasonable.

File: src/state/persist.rs (hunk 1 above)
Comment (ZER0) [reply src/state/persist.rs:39]: Agreed, but I'm not even sure this is the best way to obtain the size of a `PersistId` that derive `Canon`. Is there a better way?

Comment (krl) [reply src/state/persist.rs:39]: A comment explaining the number would also suffice imo

**Pull request merged**